z3/include/z3_sort.h

```diff
@@ -1,27 +1,69 @@
+/*********************                                                        */
+/*! \file z3_sort.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Lindsey Stuntz
+ ** This file is part of the smt-switch project.
+ ** Copyright (c) 2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file LICENSE in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Z3 implementation of AbsSort
+ **
+ **
+ **/
+
+#pragma once
+
+#include "exceptions.h"
 #include "sort.h"
+#include "utils.h"
 
 #include "z3++.h"
 
+using namespace z3;
 namespace smt {
 
-class Z3Sort : public AbsSort
-	{
- public:
-  //Z3Sort()
-  //    {};
-  Z3Sort(::z3::sort s) : sort(s) {};
-  virtual ~Z3Sort(){};
-  std::size_t hash() const override;
-  unsigned int get_width() const override;
-  Sort get_indexsort() const override;
-  Sort get_elemsort() const override;
-  std::vector<Sort> get_domain_sorts() const override;
-  Sort get_codomain_sort() const override;
-  bool compare(const Sort s) const override;
-  SortKind get_sort_kind() const override;
-
- protected:
-     ::z3::sort sort;
+// forward declaration
+class Z3Solver;
+
+class Z3Sort: public AbsSort {
+public:
+	// Non-functions
+	Z3Sort(sort z3sort, context &c) :
+			type(z3sort), is_function(false), z_func(c) {
+		ctx = &c;
+	}
+	;
+
+	// Functions
+	Z3Sort(sort z3sort, bool is_fun, func_decl zfunc, context &c) :
+			type(c), is_function(is_fun), z_func(zfunc) {
```

Comment:
Is there a scenario where you would call this constructor with `is_fun := false`? If not, it might be safer for future maintenance purposes to just drop that parameter and automatically set `is_function` to `true` when instantiating with a `func_decl`.

Reply:
You can probably remove the z3::sort being passed here, too, since you're initializing type with the context.